### ISSUE TITLE
`sys.call(1L)` vs `sys.call(-1L)`

### ISSUE THREAD

**Issue by brodieG** [mllg/checkmate#117]
Consider:
```
f1 <- function(num) assert_numeric(num)
f2 <- function(num) f1(num)
f2(letters)
## Error in f2(letters) : 
##  Assertion on 'num' failed: Must be of type 'numeric', not 'character'.
```
vs:
```
f1a <- function(num) if(!is.numeric(num)) stop("Not numeric")
f2a <- function(num) f1a(num)
f2a(letters)
## Error in f1a(num) : Not numeric
```
Notice how the latter returns `f1a`.  Obviously this is a design decision and you could make arguments for either, but it seems to me it makes more sense to report the call of the function that invokes the `assert*` functions since that is the function that is being checked.  This came up in writing a (as of yet incomplete) vignette comparing the various validation functions out there:
```
## Error in rmarkdown::render("extra/compare.Rmd"): Assertion on 'num' failed: Must be of type 'numeric', not 'character'.
```
In this case the top call is the call to `render`, which is obviously a bit confusing.  While less egregious in typical command line use, I think it could still be confusing to report an error in an inner function as affecting the call of the outer function.

I realize this is partly a matter of opinion.  Just an fyi in case you agree with me.

A simple fix would be to modify `[mstop](https://github.com/mllg/checkmate/blob/e193706bde3516a0e99ce47a2c0e2fa1c5ba3d1d/R/helper.R#L1)` to call `Sys.call(-n)` where `n` is the correct offset to get the calling function.  In our trivial example we can do it with something like this:

```
mstop2 <- function() sys.call(-1)

f1b <- function(num) if(!is.numeric(num)) mstop2()
f2b <- function(num) f1b(num)
f2b(letters)
## f1b(num)
```

**Comment by mllg** [comment]
I agree that it looks stupid in your vignette, but I guess this is more a knitr issue? I'll have a closer look if I find some spare time.

**Comment by brodieG** [comment]
I think the vignette makes it particularly obvious, but I do feel (IMO, again) it makes more sense generally for the error to be associated with the function the assertion is in, not whatever the outermost function that eventually led to the call to the function with the assertion.  For example, consider if I re-wrote the examples like:
```
f1 <- function(num) assert_numeric(num)
f2 <- function(x) f1(x)
f2(letters)
## Error in f2(letters) : 
##  Assertion on 'num' failed: Must be of type 'numeric', not 'character'.
```
Now the user is wondering "what is `num`" since as far as they are concerned based on the error message the problem is from `f2`, which only has `x` for a parameter.
